README
NAME
    Mojolicious::Plugin::MethodOverride - Simulate HTTP Verbs

VERSION
    Version 0.042

SYNOPSIS
      package My::App;
      use Mojo::Base 'Mojolicious';

      sub startup {
        my $self = shift;

        $self->plugin('MethodOverride');

        ...
      }

      1;

DESCRIPTION
    This plugin can simulate any HTTP verb (a.k.a. HTTP method) in
    environments where HTTP verbs other than GET and POST are not available.
    It uses the same approach as in many other restful web frameworks, where
    it replaces the "HTTP POST" method with a method given by an "HTTP"
    header. It is also possible to define a query parameter for the same
    purpose.

    Any token built of US-ASCII letters is accepted as a valid value for the
    HTTP verb.

CONFIGURATION
    The default HTTP header to override the "HTTP POST" method is
    "X-HTTP-Method-Override". Overriding through a query parameter is off by
    default.

    These settings can be changed in the plugin method call as demonstrated
    in the examples below:

      # Mojolicious
      $self->plugin(
        MethodOverride => {
          header => 'X-Tunneled-Method',
          param  => 'x-tunneled-method',
        }
      );

      # Mojolicious::Lite
      plugin 'MethodOverride',
        header => 'X-HTTP-Method',
        param  => 'http_method';

    HTTP header can be disabled by setting to "undef":

      # A Mojolicious app, that enables method overriding
      # by query parameter only:
      $self->plugin(
        MethodOverride => {
          header => undef,
          param  => 'x-tunneled-method',
        }
      );

AUTHOR
    Bernhard Graf "<graf at cpan.org>"

BUGS
    Please report any bugs or feature requests to
    "bug-mojolicious-plugin-methodoverride at rt.cpan.org", or through the
    web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Mojolicious-Plugin-Metho
    dOverride>. I will be notified, and then you'll automatically be
    notified of progress on your bug as I make changes.

SEE ALSO
    Plack::Middleware::MethodOverride,
    Catalyst::TraitFor::Request::REST::ForBrowsers,
    HTTP::Engine::Middleware::MethodOverride,
    <http://code.google.com/apis/gdata/docs/2.0/basics.html>

LICENSE AND COPYRIGHT
    Copyright (C) 2012 - 2015 Bernhard Graf.

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Mojolicious-Plugin-MethodOverride

0.070   2016-10-25
	- Remove the param option, because Mojolicious has the _method
	  parameter since 6.03, and to avoid issues with parameter
	  decoding.
	- Switch Makfile.PL to ExtUtils::MakeMaker
	- Use the Github issue tracker

0.060   2016-10-23
	- Fix for Mojolicious 7.0 breaking changes.

0.053   2015-06-03
	- Bug fix: t/04-override-charset.t failed w/ M<4.50.
	- Reduce requirement to Mojolicious 2.48.

0.052   2015-04-28
	- Bug fix: Clone URL query before touching it.
	- Add documentation about M:P::Charset with M:P::MethodOverride.

0.051   2015-04-19
	- Fix Version in POD.

0.050   2015-04-19
	- Override method as soon as the request header is read.
	- Mention _method parameter since Mojolicious 6.03.

0.042	2015-01-25
	- Add README (suggested by Chris Kirke).
	- Declare minimum Perl version.
	- Enable Travis CI (suggested by Chris Kirke).
	- Markdownify README.

0.041	2015-01-07
	- Remove test "method override for static files":
	  POST static_file must not be possible, but this was fixed in
	  Mojolicious 5.30. OTOH method overriding does no harm if the
	  HTTP method is completely ignored for static files before 5.30.

0.040   2014-11-23
	- No extra check for static files.

0.030   2014-05-15
	- Test fix for Mojolicious 4.91 breaking changes (chorny)

0.020   2013-05-20
	- Fix for Mojolicious 4.0 breaking changes.
	- Increase Mojolicious minimum version to 3.21.

0.010   2012-10-22
	- Accept any HTTP method name (if its made of ASCII letters).
	- Mojolicious 2 ist minimum requirement.
	- Updated test for recent Mojolicous.

0.001   2011-10-31
	First version, released on an unsuspecting world.


lib/Mojolicious/Plugin/MethodOverride.pm
package Mojolicious::Plugin::MethodOverride;

use 5.010;
use Mojo::Base 'Mojolicious::Plugin';

use Scalar::Util qw(weaken);

our $VERSION = '0.070';

sub register {
    my ($self, $app, $conf) = @_;
    my $header =
        exists $conf->{header} ? $conf->{header} : 'X-HTTP-Method-Override';

    $app->hook(
        after_build_tx => sub {
            my $tx = shift;

            weaken $tx;

            $tx->req->content->on(body => sub {
                my $req = $tx->req;

                return unless $req->method eq 'POST';

                my $method = defined $header && $req->headers->header($header);

                if ($method) {
                    $req->headers->remove($header);
                }

                if ($method and $method =~ /^[A-Za-z]+$/) {
                    $app->log->debug($header . ': ' . $method);
                    $req->method($method);
                }
            });
        }
    );
}

1;

__END__

=head1 NAME

Mojolicious::Plugin::MethodOverride - Simulate HTTP Verbs

=head1 VERSION

Version 0.070

=head1 SYNOPSIS

  package My::App;
  use Mojo::Base 'Mojolicious';

  sub startup {
    my $self = shift;

    $self->plugin(
      MethodOverride => {header => 'X-Tunneled-Method'}
    );

    ...
  }

  1;

=head1 DESCRIPTION

This plugin can simulate any HTTP verb (a.k.a. HTTP method) in
environments where HTTP verbs other than GET and POST are not available.
It uses the same approach as in many other restful web frameworks, where
it replaces the C<HTTP POST> method with a method given by an C<HTTP>
header.

Any token built of US-ASCII letters is accepted as a valid value for the
HTTP verb.

Starting with v6.03 L<Mojolicious> has a builtin parameter C<_method> to
override the HTTP method. Unfortunately there is neither a way to specify
another name, nor to disable that feature.

=head1 CONFIGURATION

The default HTTP header to override the C<HTTP POST> method is
C<"X-HTTP-Method-Override">.

These settings can be changed in the plugin method call as demonstrated
in the examples below:

  # Mojolicious
  $self->plugin(MethodOverride => {header => 'X-Tunneled-Method'});

  # Mojolicious::Lite
  plugin 'MethodOverride', {header => 'X-HTTP-Method'};

=head2 Mojolicious < 6.03

Mojolicious Versions before 6.03 (down to Mojolicious 2.48) are supported by
C<Mojolicious::Plugin::MethodOverride> version 0.060.

=head1 AUTHOR

Bernhard Graf C<< <graf at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests at
L<https://github.com/augensalat/mojolicious-plugin-methodoverride/issues>.

=head1 SEE ALSO

L<Plack::Middleware::MethodOverride>,
L<Catalyst::TraitFor::Request::REST::ForBrowsers>,
L<HTTP::Engine::Middleware::MethodOverride>,
L<http://code.google.com/apis/gdata/docs/2.0/basics.html>

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2012 - 2016 Bernhard Graf.

This library is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

See http://dev.perl.org/licenses/ for more information.
